### ISSUE TITLE
Switch to Network Manager (nmcli) for network configuration on Ubuntu

### ISSUE THREAD

**Issue by lmakarov** [docksal/docksal#321]
#319 covered a problem when DNS settings were not taking effect via `/etc/network/interfaces` (though the custom Docksal subnet was created as expected).

Using `/etc/network/interfaces` for network configuration on Ubuntu is deprecated.

Network Manager is the current way of managing network connections with a cli tool called `nmcli`
https://www.tecmint.com/configure-network-connections-using-nmcli-tool-in-linux/

Looks like `nmcli` is a standard tool across Debian family distros, so we should have the unofficial compatibility across the Debian family preserved.

**Comment by ohthehugemanatee** [comment]
ifupdown is no longer installed with Ubuntu, I think as of 17.04. If you're already using it, `/etc/network/interfaces` is respected... but for anyone who installed this year... docksal DNS doesn't work.

Techncially Ubuntu has switched to a new YML based system, but they (and it) continue to support nmcli. That's definitely the way to go for broader compatibility.

**Comment by lmakarov** [comment]
From [this](https://askubuntu.com/questions/1786/what-is-the-difference-between-network-manager-and-ifconfig-ifup-etc) thread it sounds like network-manager (and `nmcli`) is not compatible and should not be used together with the "traditional tools" (`ifconfig`, `ifup`, `ifdown` and `/etc/network/interfaces`).

While `ifup`, `ifdown` are not installed by default, neither is network-manager (tested with Ubuntu 17.10 64bit on DigitalOcean), but `ifconfig` is installed and settings in `/etc/network/interfaces` are respected, as you said.

I wonder whether the problem with the DNS settings can be simply resolved by installing the missing tools vs taking the `nmcli` route.

Testing with a fresh Ubuntu 17.10  and installing Docksal right away:

```
$ $ curl -L get.docksal.io | sh
...
Adding a subnet for Docksal...
sudo: ifup: command not found
sudo: resolvconf: command not found
...
Resetting network settings...
Resetting Docksal services...
 * proxy
/usr/bin/docker: Error response from daemon: driver failed programming external connectivity on endpoint docksal-vhost-proxy (5e331a80bac349ade9d5cd969b0938769d7ede881a9efd00253a70abf078d7b0): Error starting userland proxy: listen tcp 192.168.64.100:443: bind: cannot assign requested address.
 * dns and resolver for .docksal domain
/usr/bin/docker: Error response from daemon: driver failed programming external connectivity on endpoint docksal-dns (52e1fce13321e04c2a62ef4afe187ab4d499754af4e804a64997a48d985409bd): Error starting userland proxy: listen udp 192.168.64.100:53: bind: cannot assign requested address.
Enabling automatic *.docksal DNS resolver...
 * ssh-agent
```

Installing the missing tools first, then installing Docksal:

```
$ sudo apt-get update
...
$ sudo apt-get install ifupdown resolvconf -y
...
$ curl -L get.docksal.io | sh
...
$ newgrp docker
```

```
$ root@ubuntu-512mb-sfo2-01:~# fin docker ps
WARNING: No swap limit support
WARNING: No swap limit support
CONTAINER ID        IMAGE                     COMMAND                  CREATED              STATUS              PORTS                                                    NAMES
91708cd0fa5d        docksal/ssh-agent:1.0     "/run.sh ssh-agent"      About a minute ago   Up About a minute                                                            docksal-ssh-agent
df7b457cab33        docksal/dns:1.0           "/opt/entrypoint.sh …"   About a minute ago   Up About a minute   192.168.64.100:53->53/udp                                docksal-dns
969b79fa7e8a        docksal/vhost-proxy:1.1   "docker-entrypoint.s…"   About a minute ago   Up About a minute   192.168.64.100:80->80/tcp, 192.168.64.100:443->443/tcp   docksal-vhost-proxy
```

```
$ nslookup test.docksal
Server:		192.168.64.100
Address:	192.168.64.100#53

Name:	test.docksal
Address: 192.168.64.100
```

In other words, everything works as expected.

**Comment by ohthehugemanatee** [comment]
Ugh. Not a fan of installing requiring such a big thing as a totally new network manager for this to work.

What's wrong with just modifying /etc/resolv.conf? That works on every system.

**Comment by lmakarov** [comment]
> What's wrong with just modifying /etc/resolv.conf?

The part in all caps below

```
leonid@ubuntu-xenial-vm:~$ cat /etc/resolv.conf 
# Dynamic resolv.conf(5) file for glibc resolver(3) generated by resolvconf(8)
#     DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN
nameserver 192.168.64.100
nameserver 127.0.1.1
search localdomain
```

**Comment by lmakarov** [comment]
Closing this, as it looks like all we need on recent Ubuntu versions ([confirmed](https://github.com/docksal/docksal/issues/384#issuecomment-349012108) with 17.10) is:

```
sudo apt-get install ifupdown resolvconf -y
```

**Comment by lmakarov** [comment]
I went ahead and upgraded my local Ubuntu test VM to a fresh Ubuntu 17.10 Desktop 64bit.

Important things to note:
- Ubuntu Server and Ubuntu Desktop install different tools
- `nmcli` is available out of the box in 17.10 Desktop, but not in Server (16.04 and 17.10) (testing on DigitalOcean)
- `ifconfig` is available in Server (16.04 and 17.10), Desktop 16.04, but not in Desktop 17.10. [Reference](https://insights.ubuntu.com/2017/07/07/if-youre-still-using-ifconfig-youre-living-in-the-past/).
- `ifup/ifdown` are installed everywhere
- 17.10 (D and S) ships with [netplan](https://wiki.ubuntu.com/Netplan) - the new network management tool (stores configs in yaml). This is not available in 16.04 LTS
- `curl` in not installed in 17.10 Desktop... srsly? But `wget` is.

Looks like the only important thing that is missing on the recent Ubuntu Desktop versions is `resolvconf`. We'll still do `sudo apt-get install ifupdown resolvconf -y` during Docksal install just in case.

Long term, when `netplan` lands in 18.04 LTS. We'll consider switching to that.

**Comment by vaza18** [comment]
>sudo apt-get install ifupdown resolvconf -y
This doesn't help on Ubuntu 18.04
